src/Language/DifferentialDatalog/TypeInference.hs

```diff
@@ -0,0 +1,637 @@
+{-
+Copyright (c) 2020 VMware, Inc.
+SPDX-License-Identifier: MIT
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+-}
+
+{-# LANGUAGE FlexibleContexts, RecordWildCards, TupleSections, ImplicitParams #-}
+
+module Language.DifferentialDatalog.TypeInference(
+    TypeInferenceResult(..),
+    inferTypes
+)
+where
+
+import qualified Data.Map as M
+import Control.Monad.State.Lazy
+import Data.Maybe
+
+import Language.DifferentialDatalog.Attribute
+import Language.DifferentialDatalog.ECtx
+import Language.DifferentialDatalog.Expr
+import Language.DifferentialDatalog.Name
+import Language.DifferentialDatalog.NS
+import Language.DifferentialDatalog.Ops
+import Language.DifferentialDatalog.Rule
+import Language.DifferentialDatalog.Syntax
+import Language.DifferentialDatalog.Type
+import Language.DifferentialDatalog.Util
+import Language.DifferentialDatalog.Var
+
+data TypeInferenceResult = TypeInferenceOk (M.Map Var Type) (M.Map ECtx Type)
+                         | TypeInferenceConflict
+                         | TypeInferenceAmbiguity
+
+-- Uniquely identifies an expression in a DDlog program.
+data DDExpr = DDExpr {ddexprCtx::ECtx, ddexprExpr::Expr}
+
+-- A type constraint is an atomic predicate or a disjuncion of conjuntions of
+-- predicates.  We use a unification-based solver to resolv type constraints,
+-- which does not handle arbitrary Boolean combination constraints, hence this
+-- restricted structure.
+data Constraint = CPredicate Predicate
```

Comment:
To give nice error messages you may also want to keep with each constraint lineage information - this can be either something that implements pos, or another constraint; this way, when a constraint fails you can indicate where it originated and finally point to the source-level position that caused the constraint to be generated. Look at the llvm error messages when type unification fails for an inspiration.

Reply:
I plan to do this in the next PR. It will be clearer how to generate explanations and what state needs to be tracked as I make progress on the algorithm.